vhdbuilder/packer/trivy-scan.sh

```diff
@@ -1,41 +1,142 @@
+    exit 1
+fi
+
+if [[ "${ENABLE_TRUSTED_LAUNCH}" == "True" ]]; then
+    az login --identity --allow-no-subscriptions
+else
+    az login --identity
+fi
+
 arch="$(uname -m)"
 if [ "${arch,,}" == "arm64" ] || [ "${arch,,}" == "aarch64" ]; then
     TRIVY_ARCH="Linux-ARM64"
+    GO_ARCH="arm64"
 elif [ "${arch,,}" == "x86_64" ]; then
     TRIVY_ARCH="Linux-64bit"
+    GO_ARCH="amd64"
 else
     echo "invalid architecture ${arch,,}"
     exit 1
 fi
 
-mkdir -p "$(dirname "${TRIVY_REPORT_JSON_PATH}")"
-mkdir -p "$(dirname "${TRIVY_REPORT_TABLE_PATH}")"
+mkdir -p "$(dirname "${TRIVY_REPORT_DIRNAME}")"
 
 wget "https://github.com/aquasecurity/trivy/releases/download/v${TRIVY_VERSION}/trivy_${TRIVY_VERSION}_${TRIVY_ARCH}.tar.gz"
 tar -xvzf "trivy_${TRIVY_VERSION}_${TRIVY_ARCH}.tar.gz"
 rm "trivy_${TRIVY_VERSION}_${TRIVY_ARCH}.tar.gz"
 chmod a+x trivy 
 
-./trivy --scanners vuln rootfs -f json --skip-dirs /var/lib/containerd --ignore-unfixed --severity HIGH,CRITICAL -o "${TRIVY_REPORT_JSON_PATH}" /
+# pull vuln-to-kusto binary
+az storage blob download --auth-mode login --account-name vulntokustobinary -c vuln-to-kusto \
+    --blob-url https://vulntokustobinary.blob.core.windows.net/vuln-to-kusto/v0.0.1-5a0563d8080/vuln-to-kusto-vhd_linux_${GO_ARCH} \
+    --file ./vuln-to-kusto-vhd
+chmod a+x vuln-to-kusto-vhd
+
+# shellcheck disable=SC2155
+export PATH="$(pwd):$PATH"
+
+./trivy --scanners vuln rootfs -f json --skip-dirs /var/lib/containerd --ignore-unfixed --severity HIGH,CRITICAL -o "${TRIVY_REPORT_ROOTFS_JSON_PATH}" /
+./vuln-to-kusto-vhd scan-report \
+    --vhd-vhdname="${VHD_NAME}" \
+    --vhd-ossku="${OS_SKU}" \
+    --vhd-osversion="${OS_VERSION}" \
+    --vhd-skuname="${SKU_NAME}" \
+    --kusto-endpoint=${KUSTO_ENDPOINT} \
+    --kusto-database=${KUSTO_DATABASE} \
+    --kusto-table=${KUSTO_TABLE} \
+    ${TRIVY_REPORT_ROOTFS_JSON_PATH}
 
 IMAGE_LIST=$(ctr -n k8s.io image list -q | grep -v sha256)
 
 echo "This contains the list of images with high and critical level CVEs (if present), that are present in the node. 
-Note: images without CVEs are also listed" >> "${TRIVY_REPORT_TABLE_PATH}"
+Note: images without CVEs are also listed" >> "${TRIVY_REPORT_IMAGE_TABLE_PATH}"
+
+for CONTAINER_IMAGE in $IMAGE_LIST; do
+    # append to table
+    ./trivy --scanners vuln image --ignore-unfixed --severity HIGH,CRITICAL -f table ${CONTAINER_IMAGE} >> ${TRIVY_REPORT_IMAGE_TABLE_PATH} || true
 
-for image in $IMAGE_LIST; do
-    ./trivy --scanners vuln image --ignore-unfixed --severity HIGH,CRITICAL -f table $image >> ${TRIVY_REPORT_TABLE_PATH} || true
+    # export to Kusto, one by one
+    BASE_CONTAINER_IMAGE=$(basename ${CONTAINER_IMAGE})
+    TRIVY_REPORT_IMAGE_JSON_PATH=${TRIVY_REPORT_DIRNAME}/trivy-report-image-${BASE_CONTAINER_IMAGE}.json
+    ./trivy --scanners vuln image -f json --ignore-unfixed --severity HIGH,CRITICAL -o ${TRIVY_REPORT_IMAGE_JSON_PATH} $CONTAINER_IMAGE || true
```

Comment:
why are we `|| true` , ignoring failure with the scanner ?

Reply:
I believe this was for skipping over images that trivy fails to scan
"
./vuln-to-kusto-vhd scan-report --vhd-vhdname=2204-gen2-containerd --vhd-ossku=Ubuntu --vhd-osversion=22.04 --vhd-skuname=2204gen2containerd --kusto-endpoint=https://aksdevinfra.westus2.kusto.windows.net/ --kusto-database=AKSShieldGuard --kusto-table=AKSContainerCVEScanner /opt/azure/containers/trivy-report-image-aks-node-ca-watcher:master.240530.2.json\n2024/08/08 01:56:13 WARN no environment resolved, fallback to unknown environment component=environment-detector\n2024/08/08 01:56:13 WARN no build source detected, fallback to unknown build source component=build-source-detector\n+ rm /opt/azure/containers/trivy-report-image-aks-node-ca-watcher:master.240530.2.json\n+ for CONTAINER_IMAGE in $IMAGE_LIST\n+ ./trivy --scanners vuln image --ignore-unfixed --severity HIGH,CRITICAL -f table mcr.microsoft.com/aks/aks-node-ca-watcher:static\n2024-08-08T01:56:14.311Z\t\u001b[34mINFO\u001b[0m\tVulnerability scanning is enabled\n2024-08-08T01:56:14.477Z\t\u001b[31mFATAL\u001b[0m\timage scan error: scan error: unable to initialize a scanner: unable to initialize a docker scanner: 4 errors occurred:\n\t* unable to inspect the image (mcr.microsoft.com/aks/aks-node-ca-watcher:static): Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?\n\t* unable to initialize Podman client: no podman socket found: stat podman/podman.sock: no such file or directory\n\t* image not found in containerd store: mcr.microsoft.com/aks/aks-node-ca-watcher:static\n\t* GET https://mcr.microsoft.com/v2/aks/aks-node-ca-watcher/manifests/static: MANIFEST_UNKNOWN: manifest tagged by \"static\" is not found;
"